src/datamodel/builder/UMessage.cpp

```diff
@@ -10,3 +10,236 @@
+    v1::UUri sink = request.attributes().source();
+    v1::UUID reqId = request.attributes().reqid();
+    v1::UPriority priority = request.attributes().priority();
+    v1::UUri method = request.attributes().sink();
+
+    return UMessageBuilder::response(std::move(sink), std::move(reqId),
+                                      priority, std::move(method));
+
+}
+
+UMessageBuilder& UMessageBuilder::withPriority(v1::UPriority priority) {
+    if ((priority < v1::UPriority_MIN) ||
+        (priority > v1::UPriority_MAX)) {
+        throw std::out_of_range("Priority value is out of range");
+    }
+
+    if ((attributes_.type() == v1::UMessageType::UMESSAGE_TYPE_REQUEST) ||
+        (attributes_.type() == v1::UMessageType::UMESSAGE_TYPE_RESPONSE)) {
+        if (priority < v1::UPriority::UPRIORITY_CS4) {
+            throw std::out_of_range("Priority value is out of range, for req/resp should be > CS4");
+        }
+    }
+
+    attributes_.set_priority(priority);
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withTtl(std::chrono::milliseconds ttl) {
+    if ((ttl.count() <= 0) || (ttl.count() > std::numeric_limits<uint32_t>::max())) {
+        throw std::out_of_range("TTL value is out of range");
+    }
+
+    attributes_.set_ttl(static_cast<uint32_t>(ttl.count()));
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withToken(const std::string& token) {
+    if ((attributes_.type() != v1::UMessageType::UMESSAGE_TYPE_REQUEST)) {
+        throw std::domain_error("Token can only be set on a request message");
+    }
+
+    attributes_.set_token(token);
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withPermissionLevel(uint32_t level) {
+    if ((attributes_.type() != v1::UMessageType::UMESSAGE_TYPE_REQUEST)) {
+        throw std::domain_error("Permission level can only be set on a request message");
+    }
+
+    attributes_.set_permission_level(level);
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withCommStatus(v1::UCode code) {
+    if (attributes_.type() != v1::UMessageType::UMESSAGE_TYPE_RESPONSE) {
+        throw std::domain_error("CommStatus can only be set on a response message");
+    }
+
+    if ( ! (v1::UCode_IsValid(code)) ) {
+        throw std::out_of_range("UCode is out of range" + std::to_string(code));
+    }
+
+    if (code != v1::UCode::OK) {
+        attributes_.set_commstatus(code);
+    }
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withPayloadFormat(v1::UPayloadFormat format) {
+    if ((format < v1::UPayloadFormat_MIN) ||
+        (format > v1::UPayloadFormat_MAX)) {
+        throw std::out_of_range("Payload format value is out of range");
+    }
+
+    attributes_.set_payload_format(format);
+
+    return *this;
+}
+
+v1::UMessage UMessageBuilder::build() const {
+    v1::UMessage message;
+    if ((attributes_.payload_format() >= v1::UPayloadFormat_MIN) &&
+        (attributes_.payload_format() <= v1::UPayloadFormat_MAX)) {
```

Comment:
This is where storing the intermediate values in a UAttributes poses a challenge. It would be better if there was a `std::optional<v1::UPayloadFormat> format_` in the class. Then it's easy to tell if it has ever been set (`if (format_)...`).

Reply:
Done